File: runtime/src/main/java/scooterdelta/challenge/bot/ChallengeApplication.java
package scooterdelta.challenge.bot;

import dagger.ObjectGraph;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scooterdelta.challenge.bot.module.ChallengeModule;
import scooterdelta.challenge.bot.process.ProcessEngine;
import scooterdelta.challenge.bot.process.module.ProcessModule;

import javax.inject.Inject;
import java.io.File;

import static dagger.ObjectGraph.create;

public class ChallengeApplication implements Runnable {

    private final ProcessEngine processEngine;

    private static final Logger LOGGER = LoggerFactory.getLogger(ChallengeApplication.class);

    @Inject
    public ChallengeApplication(final ProcessEngine processEngine) {
        this.processEngine = processEngine;
    }

    @Override
    public void run() {
        processEngine.run();
    }

    public static void main(final String[] args) {

        final long startTime = System.nanoTime();
        runBot(args);
        final long endTime = System.nanoTime();

        LOGGER.debug(String.format("Bot finished in %d ms.", (endTime - startTime) / 1000000));
    }

    private static void runBot(final String... args) {

        if (args.length != 2) {
            printUsage();
            System.exit(1);
        }

        final String key = args[0];
        final File workingDir = new File(args[1]);

        if (!workingDir.exists()) {
            printUsage();
            LOGGER.debug(String.format("Error: Working directory \" %s \" does not exist.", args[1]));
            System.exit(1);
        }

        final ObjectGraph objectGraph = create(
                new ChallengeModule(),
                new ProcessModule(workingDir, key)
        );
        final ChallengeApplication challengeApplication = objectGraph.get(ChallengeApplication.class);

        challengeApplication.run();
    }

    private static void printUsage() {

        LOGGER.debug("Java SampleBot usage: sample-bot.jar <PlayerKey> <WorkingDirectoryFilename>");
        LOGGER.debug("\tPlayerKey\tThe key assigned to your bot.");
        LOGGER.debug("\tWorkingDirectoryFilename\tThe working directory folder where the match runner will output map and state files and look for the move file.");
    }
}

File: runtime/src/main/kotlin/scooterdelta/challenge/bot/ChallengeApplication.kt
package scooterdelta.challenge.bot

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import scooterdelta.challenge.bot.component.ChallengeComponent
import scooterdelta.challenge.bot.component.DaggerChallengeComponent
import scooterdelta.challenge.bot.process.module.ProcessModule
import java.io.File

object ChallengeApplication {

    val LOGGER : Logger = LoggerFactory.getLogger(ChallengeApplication.javaClass)

    @JvmStatic
    fun main(args: Array<String>) {
        val startTime : Long = System.nanoTime()
        runBot(args)
        val endTime : Long = System.nanoTime()

        LOGGER.debug(String.format("Bot finished in %d ms.", (endTime - startTime) / 1000000))
    }

    fun runBot(args: Array<String>) {

        if (args.size != 2) {
            printError()
            System.exit(1)
        }

        val key : String = args[0]
        val workingDir : File = File(args[1])

        if (!workingDir.exists()) {
            printError()
            LOGGER.error("Error: Working directory \" {} \" does not exist.", args[1])
            System.exit(1)
        }

        val challengeComponent : ChallengeComponent = DaggerChallengeComponent.builder()
                .processModule(ProcessModule(workingDir, key))
                .build()

        // Run the application
        challengeComponent.getChallengeRunner().run()
    }

    fun printError() {
        LOGGER.error("Java SampleBot usage: sample-bot.jar <PlayerKey> <WorkingDirectoryFilename>")
        LOGGER.error("\tPlayerKey\tThe key assigned to your bot.")
        LOGGER.error("\tWorkingDirectoryFilename\tThe working directory folder where the match runner will output map and state files and look for the move file.")
    }
}
